Support `kustomize edit add base`

Add `kustomize edit add base` command to modify the kustomization.yaml with a base.

was https://github.com/kubernetes/kubectl/issues/481

@Liujingfang1 actually if you could possibly explain the steps on how to build the binary locally?
Running `go build kustomize.go` on master gives me the following error when I try executing the binary:
```
@guin:kustomize (master) 🦉  go build kustomize.go 
@guin:kustomize (master %) 🦉  ./kustomize
./kustomize flag redefined: log_dir
panic: ./kustomize flag redefined: log_dir

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc420092060, 0x1c249a0, 0xc420030670, 0x1b52723, 0x7, 0x1b74c2c, 0x2f)
	/usr/local/Cellar/go/1.10.2/libexec/src/flag/flag.go:810 +0x540
flag.(*FlagSet).StringVar(0xc420092060, 0xc420030670, 0x1b52723, 0x7, 0x0, 0x0, 0x1b74c2c, 0x2f)
	/usr/local/Cellar/go/1.10.2/libexec/src/flag/flag.go:713 +0x8b
flag.(*FlagSet).String(0xc420092060, 0x1b52723, 0x7, 0x0, 0x0, 0x1b74c2c, 0x2f, 0xc420030660)
	/usr/local/Cellar/go/1.10.2/libexec/src/flag/flag.go:726 +0x8b
flag.String(0x1b52723, 0x7, 0x0, 0x0, 0x1b74c2c, 0x2f, 0x11f0c61)
	/usr/local/Cellar/go/1.10.2/libexec/src/flag/flag.go:733 +0x69
```

--- a/pkg/commands/addbase.go
+++ b/pkg/commands/addbase.go
@@ -1,0 +1,95 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/kubernetes-sigs/kustomize/pkg/constants"
+	"github.com/kubernetes-sigs/kustomize/pkg/fs"
+)
+
+type addBaseOptions struct {
+	baseDirectoryPath string
+}
+
+// newCmdAddBase adds the file path of the kustomize base to the kustomization file.
+func newCmdAddBase(fsys fs.FileSystem) *cobra.Command {
+	var o addBaseOptions
+
+	cmd := &cobra.Command{
+		Use:   "base",
+		Short: "Adds a directory path to a base kustomization to the current directory's kustomization file.",
+		Example: `
+		add base {filepath}`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := o.Validate(args)
+			if err != nil {
+				return err
+			}
+			err = o.Complete(cmd, args)
+			if err != nil {
+				return err
+			}
+			return o.RunAddBase(fsys)
+		},
+	}
+	return cmd
+}
+
+// Validate validates addBase command.
+func (o *addBaseOptions) Validate(args []string) error {
+	if len(args) != 1 {
+		return errors.New("must specify a base directory")
+	}
+	o.baseDirectoryPath = args[0]
+	return nil
+}
+
+// Complete completes addBase command.
+func (o *addBaseOptions) Complete(cmd *cobra.Command, args []string) error {
+	return nil
+}
+
+// RunAddBase runs addBase command (do real work).
+func (o *addBaseOptions) RunAddBase(fsys fs.FileSystem) error {
+	_, err := fsys.Stat(o.baseDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	mf, err := newKustomizationFile(constants.KustomizationFileName, fsys)
+	if err != nil {
+		return err
+	}
+
+	m, err := mf.read()
+	if err != nil {
+		return err
+	}
+
+	if stringInSlice(o.baseDirectoryPath, m.Bases) {
+		return fmt.Errorf("base %s already in kustomization file", o.baseDirectoryPath)
+	}
+
+	m.Bases = append(m.Bases, o.baseDirectoryPath)
+
+	return mf.write(m)
+}

--- a/pkg/commands/addbase.go
+++ b/pkg/commands/addbase.go
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
@@ -19,6 +19,7 @@ package commands
 import (
 	"errors"
 	"fmt"
+	"strings"
 
 	"github.com/spf13/cobra"
 
@@ -27,7 +28,7 @@ import (
 )
 
 type addBaseOptions struct {
-	baseDirectoryPath string
+	baseDirectoryPaths string
 }
 
 // newCmdAddBase adds the file path of the kustomize base to the kustomization file.
@@ -36,9 +37,9 @@ func newCmdAddBase(fsys fs.FileSystem) *cobra.Command {
 
 	cmd := &cobra.Command{
 		Use:   "base",
-		Short: "Adds a directory path to a base kustomization to the current directory's kustomization file.",
+		Short: "Adds one or more bases to the kustomization.yaml in current directory",
 		Example: `
-		add base {filepath}`,
+		add base {filepath1},{filepath2}`,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			err := o.Validate(args)
 			if err != nil {
@@ -59,7 +60,7 @@ func (o *addBaseOptions) Validate(args []string) error {
 	if len(args) != 1 {
 		return errors.New("must specify a base directory")
 	}
-	o.baseDirectoryPath = args[0]
+	o.baseDirectoryPaths = args[0]
 	return nil
 }
 
@@ -70,11 +71,6 @@ func (o *addBaseOptions) Complete(cmd *cobra.Command, args []string) error {
 
 // RunAddBase runs addBase command (do real work).
 func (o *addBaseOptions) RunAddBase(fsys fs.FileSystem) error {
-	_, err := fsys.Stat(o.baseDirectoryPath)
-	if err != nil {
-		return err
-	}
-
 	mf, err := newKustomizationFile(constants.KustomizationFileName, fsys)
 	if err != nil {
 		return err
@@ -85,11 +81,19 @@ func (o *addBaseOptions) RunAddBase(fsys fs.FileSystem) error {
 		return err
 	}
 
-	if stringInSlice(o.baseDirectoryPath, m.Bases) {
-		return fmt.Errorf("base %s already in kustomization file", o.baseDirectoryPath)
-	}
+	// split directory paths
+	paths := strings.Split(o.baseDirectoryPaths, ",")
+	for _, path := range paths {
+		_, err := fsys.Stat(path)
+		if err != nil {
+			return err
+		}
+		if stringInSlice(path, m.Bases) {
+			return fmt.Errorf("base %s already in kustomization file", path)
+		}
+		m.Bases = append(m.Bases, path)
 
-	m.Bases = append(m.Bases, o.baseDirectoryPath)
+	}
 
 	return mf.write(m)
 }

--- a/pkg/commands/addbase_test.go
+++ b/pkg/commands/addbase_test.go
@@ -1,0 +1,85 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"testing"
+
+	"strings"
+
+	"github.com/kubernetes-sigs/kustomize/pkg/constants"
+	"github.com/kubernetes-sigs/kustomize/pkg/fs"
+)
+
+const (
+	baseDirectoryPath = "my/path/to/wonderful/base"
+)
+
+func TestAddBaseHappyPath(t *testing.T) {
+	fakeFS := fs.MakeFakeFS()
+	fakeFS.Mkdir(baseDirectoryPath, 0777)
+	fakeFS.WriteFile(constants.KustomizationFileName, []byte(kustomizationContent))
+
+	cmd := newCmdAddBase(fakeFS)
+	args := []string{baseDirectoryPath}
+	err := cmd.RunE(cmd, args)
+	if err != nil {
+		t.Errorf("unexpected cmd error: %v", err)
+	}
+	content, err := fakeFS.ReadFile(constants.KustomizationFileName)
+	if err != nil {
+		t.Errorf("unexpected read error: %v", err)
+	}
+	if !strings.Contains(string(content), baseDirectoryPath) {
+		t.Errorf("expected patch name in kustomization")
+	}
+}
+
+func TestAddBaseAlreadyThere(t *testing.T) {
+	fakeFS := fs.MakeFakeFS()
+	// Create fake directory
+	fakeFS.Mkdir(baseDirectoryPath, 0777)
+	fakeFS.WriteFile(constants.KustomizationFileName, []byte(kustomizationContent))
+
+	cmd := newCmdAddBase(fakeFS)
+	args := []string{baseDirectoryPath}
+	err := cmd.RunE(cmd, args)
+	if err != nil {
+		t.Fatalf("unexpected cmd error: %v", err)
+	}
+	// adding an existing base should return an error
+	err = cmd.RunE(cmd, args)
+	if err == nil {
+		t.Errorf("expected already there problem")
+	}
+	if err.Error() != "base "+baseDirectoryPath+" already in kustomization file" {
+		t.Errorf("unexpected error %v", err)
+	}
+}
+
+func TestAddBaseNoArgs(t *testing.T) {
+	fakeFS := fs.MakeFakeFS()
+
+	cmd := newCmdAddBase(fakeFS)
+	err := cmd.Execute()
+	if err == nil {
+		t.Errorf("expected error: %v", err)
+	}
+	if err.Error() != "must specify a base directory" {
+		t.Errorf("incorrect error: %v", err.Error())
+	}
+}

--- a/pkg/commands/addbase_test.go
+++ b/pkg/commands/addbase_test.go
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
@@ -26,16 +26,19 @@ import (
 )
 
 const (
-	baseDirectoryPath = "my/path/to/wonderful/base"
+	baseDirectoryPaths = "my/path/to/wonderful/base,other/path/to/even/more/wonderful/base"
 )
 
 func TestAddBaseHappyPath(t *testing.T) {
 	fakeFS := fs.MakeFakeFS()
-	fakeFS.Mkdir(baseDirectoryPath, 0777)
+	bases := strings.Split(baseDirectoryPaths, ",")
+	for _, base := range bases {
+		fakeFS.Mkdir(base, 0777)
+	}
 	fakeFS.WriteFile(constants.KustomizationFileName, []byte(kustomizationContent))
 
 	cmd := newCmdAddBase(fakeFS)
-	args := []string{baseDirectoryPath}
+	args := []string{baseDirectoryPaths}
 	err := cmd.RunE(cmd, args)
 	if err != nil {
 		t.Errorf("unexpected cmd error: %v", err)
@@ -44,19 +47,25 @@ func TestAddBaseHappyPath(t *testing.T) {
 	if err != nil {
 		t.Errorf("unexpected read error: %v", err)
 	}
-	if !strings.Contains(string(content), baseDirectoryPath) {
-		t.Errorf("expected patch name in kustomization")
+
+	for _, base := range bases {
+		if !strings.Contains(string(content), base) {
+			t.Errorf("expected base name in kustomization")
+		}
 	}
 }
 
 func TestAddBaseAlreadyThere(t *testing.T) {
 	fakeFS := fs.MakeFakeFS()
-	// Create fake directory
-	fakeFS.Mkdir(baseDirectoryPath, 0777)
+	// Create fake directories
+	bases := strings.Split(baseDirectoryPaths, ",")
+	for _, base := range bases {
+		fakeFS.Mkdir(base, 0777)
+	}
 	fakeFS.WriteFile(constants.KustomizationFileName, []byte(kustomizationContent))
 
 	cmd := newCmdAddBase(fakeFS)
-	args := []string{baseDirectoryPath}
+	args := []string{baseDirectoryPaths}
 	err := cmd.RunE(cmd, args)
 	if err != nil {
 		t.Fatalf("unexpected cmd error: %v", err)
@@ -66,9 +75,15 @@ func TestAddBaseAlreadyThere(t *testing.T) {
 	if err == nil {
 		t.Errorf("expected already there problem")
 	}
-	if err.Error() != "base "+baseDirectoryPath+" already in kustomization file" {
-		t.Errorf("unexpected error %v", err)
+	var expectedErrors []string
+	for _, base := range bases {
+		error := "base " + base + " already in kustomization file"
+		expectedErrors = append(expectedErrors, error)
+		if !stringInSlice(error, expectedErrors) {
+			t.Errorf("unexpected error %v", err)
+		}
 	}
+
 }
 
 func TestAddBaseNoArgs(t *testing.T) {

--- a/pkg/commands/commands.go
+++ b/pkg/commands/commands.go
@@ -81,7 +81,7 @@ func newCmdEdit(fsys fs.FileSystem) *cobra.Command {
 func newCmdAdd(fsys fs.FileSystem) *cobra.Command {
 	c := &cobra.Command{
 		Use:   "add",
-		Short: "Adds configmap/resource/patch to the kustomization file.",
+		Short: "Adds configmap/resource/patch/base to the kustomization file.",
 		Long:  "",
 		Example: `
 	# Adds a configmap to the kustomization file
@@ -92,6 +92,9 @@ func newCmdAdd(fsys fs.FileSystem) *cobra.Command {
 
 	# Adds a patch to the kustomization
 	kustomize edit add patch <filepath>
+
+	# Adds a base directory to the kustomization
+	kustomize edit add base <filepath>
 `,
 		Args: cobra.MinimumNArgs(1),
 	}
@@ -99,6 +102,7 @@ func newCmdAdd(fsys fs.FileSystem) *cobra.Command {
 		newCmdAddResource(fsys),
 		newCmdAddPatch(fsys),
 		newCmdAddConfigMap(fsys),
+		newCmdAddBase(fsys),
 	)
 	return c
 }

--- a/pkg/commands/commands.go
+++ b/pkg/commands/commands.go
@@ -93,8 +93,9 @@ func newCmdAdd(fsys fs.FileSystem) *cobra.Command {
 	# Adds a patch to the kustomization
 	kustomize edit add patch <filepath>
 
-	# Adds a base directory to the kustomization
+	# Adds one or more base directories to the kustomization
 	kustomize edit add base <filepath>
+	kustomize edit add base <filepath1>,<filepath2>,<filepath3>
 `,
 		Args: cobra.MinimumNArgs(1),
 	}

